src/Npgsql/NpgsqlConnection.cs

```diff
@@ -1880,32 +1909,49 @@ public void ReloadTypes()
         CheckReady();
 
         using var scope = StartTemporaryBindingScope(out var connector);
-
-        _dataSource!.Bootstrap(
-            connector,
-            NpgsqlTimeout.Infinite,
-            forceReload: true,
-            async: false,
-            CancellationToken.None)
-            .GetAwaiter().GetResult();
+        BootstrapDataSource(connector, async: false).GetAwaiter().GetResult();
     }
 
     /// <summary>
     /// Flushes the type cache for this connection's connection string and reloads the types for this connection only.
     /// Type changes will appear for other connections only after they are re-opened from the pool.
     /// </summary>
-    public async Task ReloadTypesAsync()
+    public Task ReloadTypesAsync()
     {
+        using var _ = NoSynchronizationContextScope.Enter();
         CheckReady();
 
         using var scope = StartTemporaryBindingScope(out var connector);
+        return BootstrapDataSource(connector, async: true);
+    }
 
-        await _dataSource!.Bootstrap(
-                connector,
-                NpgsqlTimeout.Infinite,
-                forceReload: true,
-                async: true,
-                CancellationToken.None);
+    async Task BootstrapDataSource(NpgsqlConnector connector, bool async)
+    {
+        while (true)
+        {
+            try
+            {
+                await _dataSource!.Bootstrap(
+                    connector,
+                    NpgsqlTimeout.Infinite,
+                    forceReload: true,
+                    async,
+                    CancellationToken.None);
+                break;
+            }
+            // When using PoolManager (legacy code path setting NpgsqlConnection.ConnectionString), clearing a pool
+            // removes the pool (NpgsqlDataSource) from the PoolManager and disposes it.
+            // This leads to an ObjectDisposedException when calling NpgsqlDataSource.Get() here.
```

Comment:
This is for making ReloadTypes work after Clear/ClearAll is called, when multiplexing is on, right (good catch)? If so maybe mention multiplexing in the comment to make this clearer.

Reply:
That's the simplest solution to make things work if you call ReloadTypes *after* opening the first multiplexing connection.
Without it, you fail in the next command's execution after calling it and from there it it very hard to recover the DataSource.